File: 0070-climbing-stairs/0070-climbing-stairs.cpp
class Solution {
public:
    int dp[46];
    int climbStairs1(int n) {
        if(n==0){
            return 1;
        }
        if(n < 0) return 0;  
        if(dp[n]!=-1){
            return dp[n];
        }

        int oneStep = climbStairs1(n-1);
        int twoStep = climbStairs1(n-2);

        return dp[n] = oneStep+twoStep;
    }
    int climbStairs(int n){
        memset(dp, -1, sizeof(dp));
        return climbStairs1(n);
    }
};

File: 0070-climbing-stairs/0070-climbing-stairs.java
class Solution {
    public int print(int n,int []dp){
        if(n==1){
            return 1;
        }
        if(n==2){
            return 2;
        }
        if(dp[n]!=-1){
            return dp[n];
        }
        return dp[n]=print(n-1,dp)+print(n-2,dp);
    }
    public int climbStairs(int n) {
        int dp[]=new int[n+1];
        Arrays.fill(dp,-1);
        return print(n,dp);
    }
}